File: courses/codes/WebNavigationSolutions/browser.py
IGNORED = "Ignored"
forward_stack = []
backward_stack = []
current_page = "http://www.acm.org"


def do_back():
    global current_page
    if len(backward_stack) > 0:
        forward_stack.append(current_page)
        current_page = backward_stack.pop()
        print(current_page)
    else:
        print(IGNORED)


def do_forward():
    global current_page
    if len(forward_stack) > 0:
        backward_stack.append(current_page)
        current_page = forward_stack.pop()
        print(current_page)
    else:
        print(IGNORED)


def do_visit(url):
    global current_page
    backward_stack.append(current_page)
    current_page = url
    print(current_page)
    forward_stack.clear()


while True:
    cmd = input()
    if cmd == "BACK":
        do_back()
    elif cmd == "FORWARD":
        do_forward()
    elif cmd[:5] == "VISIT":
        do_visit(cmd[6:])
    elif cmd == "QUIT":
        break
    else:
        print("Invalid command")

File: courses/codes/WebNavigationSolutions/browser.cpp
// CC1SrLesson2Q1.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>
#include <string>
#include <stack>
using namespace std;

const string IGNORED = "Ignored";
stack<string> forward_stack;
stack<string> backward_stack;
string current_page = "http://www.acm.org";

void doBack() {
    if (!backward_stack.empty()) {
        forward_stack.push(current_page);
        current_page = backward_stack.top();
        backward_stack.pop();
        cout << current_page << endl;
    } else {
        cout << IGNORED << endl;
    }
}

void doForward() {
    if (!forward_stack.empty()) {
        backward_stack.push(current_page);
        current_page = forward_stack.top();
        forward_stack.pop();
        cout << current_page << endl;
    } else {
        cout << IGNORED << endl;
    }
}

void doVisit(string url) {
    backward_stack.push(current_page);
    current_page = url;
    cout << current_page << endl;
    forward_stack = stack<string>();
}

int main() {
    while (true) {
        string cmd;
        getline(cin, cmd);
        if (cmd == "BACK") {
            doBack();
        } else if (cmd == "FORWARD") {
            doForward();
        } else if (cmd.rfind("VISIT", 0) == 0) {
            doVisit(cmd.substr(6));
        } else if (cmd == "QUIT") {
            break;
        } else {
            cout << "Invalid command" << endl;
        }
    }
}
